### DISCUSSION TITLE
Is there any example or way I can consume a CsWinRT library from a CS app?

### DISCUSSION THREAD

**Discussion by BradChase2011** [microsoft/CsWinRT#1427] [Q&A]
I have a C++ UWP app example I setup that consumes a .net6+ C# library using CsWinRT.  I then setup a UWP C# app to consume the .net6+ C# library using CsWinRT and it fails when I try to access the same class with an error:

Exception thrown: 'System.Runtime.InteropServices.COMException' in BlankAppCSharp1.exe
An exception of type 'System.Runtime.InteropServices.COMException' occurred in BlankAppCSharp1.exe but was not handled in user code
Error in the DLL (Exception from HRESULT: 0x800401F9 (CO_E_ERRORINDLL))

So I guess my questions are, what might be wrong?  Are there any examples around that show how to use a C# UWP referencing a CsWinRT winmd or is that not possible?  The C++ UWP app works just fine...

Thanks much!

**Comment by manodasanW** [comment]
It sounds like you are trying to activate a CsWinRT authored .NET 6+ WinRT component in a UWP app.  I assume the UWP app is .NET native.  If so, since you are trying to load the .NET 6 runtime within a .NET native app with its own older version of .NET, that is probably going to run into issues and won't work.  Given C++ UWP has no .NET version loaded already, it doesn't run into any errors.

**Reply by BradChase2011** [reply]
Yea that is basically on point.  I dont know much about the under workings of how the code starts separating once it hits the winrt side of things.  A side note and more explanation, we are testing if we can get winrt running on an xbox using the community program.  Is this possible?  While the UWP C++ runs on the windows machine, when deployed to the xbox and we try to hit a class from the C# .net6 winrt side, we get this error:

A fatal error occurred, the default install location cannot be obtained.Exception thrown at 0x00007FF8E9523EEC (KernelBase.dll) in BlankApp1.exe: WinRT originate error - 0x80008083 : 'One of the dependent libraries is missing.'.
Exception thrown at 0x00007FF8E9523EEC in BlankApp1.exe: Microsoft C++ exception: winrt::hresult_error at memory location 0x00000021F3FFC080.

Is this a fruitless endeavor, impossible?  Or do we need to package it differently to push to the xbox?  Unfortunately the old .NET with UWP is extremely slow compared to our engine running in .NET 8 (about a tenth the speed) so we wanted to play around and see if we could get our engine running using winrt with higher performance using the community program.  If this is off topic, I can start a new discussion, or if its impossible then maybe stepping away from it.

Thanks much!

**Comment by driver1998** [comment]
Speaking of Xbox, maybe try moving the whole app to .NET 8? https://github.com/driver1998/NetCoreApp

The XAML stuff is hacked together (see https://github.com/microsoft/CsWinRT/pull/1421 for official support), but other stuff works out of the box.

**Reply by BradChase2011** [reply]
We have our own multi threaded xaml.engine, so we prefer not to have any Microsoft xaml at all.  Problem with Xbox community is it's very picky on what can and cannot run.  We were never able to get winrt running even under cpp to c#.  We did create a wrapper that wrapped our entry points into .net 8 directx side using NativeAOT and it's 4x faster than using UWP and netstandard2.  On an Xbox series s using community (about half the CPU and GPU), using UWP and netstandard2 we marked 115fps, using UWP wrapping .net 8 with NativeAOT, we marked 420fps on the same game demo.  BIG difference in performance.

**Reply by driver1998** [reply]
> Problem with Xbox community is it's very picky on what can and cannot run. We were never able to get winrt running even under cpp to c#.

Not sure about Xbox but the above sample works just fine on IoT Core, which is an equally restricted platform.

Granted I tested it before I added WinUI 2 to the mix, maybe the current master won't work.

**Reply by BradChase2011** [reply]
Yea WinUI wont run under xbox unfortunately.  Microsoft cancelled UWP and never replaced it on the xbox, hah gotta love it.  Thats the whole reason for going down this rabbit hole.  We have spent alot of time getting WinRT to run correctly and it is a massive time sink to get anything moving forward.  At this point its hard enough to get a simple templated app sample running at all, even with cpp -> cs.  I put up a bug and will see if they get that fixed and once we know it can actually work with a test library and a new app, then maybe revisit.
